src/array/cpu/labor_pick.h

```diff
@@ -0,0 +1,347 @@
+          continue;
+
+        const auto k = std::min(num_picks, d);
+
+        if (hop_map.empty()) {  // weights first iter, pi = A
+          for (auto j = indptr[rid]; j < indptr[rid + 1]; j++)
+            ps[j - indptr[rid]] = A[j];
+        } else {
+          for (auto j = indptr[rid]; j < indptr[rid + 1]; j++)
+            ps[j - indptr[rid]] = hop_map[indices[j]];
+        }
+
+        double var_target = ds[i] * ds[i] / k;
+        if (weights) {
+          var_target -= ds[i] * ds[i] / d;
+          for (auto j = indptr[rid]; j < indptr[rid + 1]; j++)
+            var_target += A[j] * A[j];
+        }
+        FloatType c = cs[i];
+        double var_1;
+        if (linear_c_convergence) {
+          // fix weights case
+          std::make_heap(ps, ps + d);
+          double r = 0;
+          if (weights) {
+            for (IdxType i = 0; i < d; i++) {
+              const auto w = A[i + indptr[rid]];
+              r += w / ps[i];
+            }
+          } else {
+            for (IdxType i = 0; i < d; i++)
+              r += 1 / ps[i];
+          }
+          c = r / var_target;
+          if (weights) {
+            double v = A[d - 1] * A[d - 1];
+            for (auto i = d - 1; i > 0 && c * ps[i] >= 1; i--) {
+              // fix this
+              r -= ps[i];
+              c = r / (var_target - v);
+              v += A[i - 1 + indptr[rid]];
+            }
+          } else {
+            for (auto i = d - 1; i > 0 && c * ps[i] >= 1; i--) {
+              std::pop_heap(ps, ps + i + 1);
+              r -= ps[i];
+              c = r / (var_target - (d - i));
+            }
+          }
+        } else {
+          do {
+            var_1 = 0;
+            if (weights) {
+              for (auto j = indptr[rid]; j < indptr[rid + 1]; j++)
+                var_1 += A[j] * A[j] / std::min(ONE, c * ps[j - indptr[rid]]);
+            } else {
+              for (auto j = indptr[rid]; j < indptr[rid + 1]; j++)
+                var_1 += ONE / std::min(ONE, c * ps[j - indptr[rid]]);
+            }
+
+            c *= var_1 / var_target;
+          } while (std::min(var_1, var_target) / std::max(var_1, var_target) < 1 - eps);
+        }
+
+        cs[i] = c;
+      }
+
+      if (!weights || iters) {
+        double cur_ex_nodes = 0;
+        for (auto it : hop_map)
+          cur_ex_nodes += std::min((FloatType)1, it.second);
+        if (cur_ex_nodes / prev_ex_nodes >= 1 - eps)
+          break;
+        prev_ex_nodes = cur_ex_nodes;
+      }
+    }
+  }
+
+  // When run distributed, random_seed needs to be equivalent.
+  auto seeds = static_cast<const int64_t *>(random_seed->data);
+  const auto num_seeds = random_seed->shape[0];
+  auto cnts = static_cast<const int64_t *>(cnt->data);
+  const auto l = (cnts[0] % cnts[1]) * 1.0 / cnts[1];
+  const auto pi = std::acos(-1.0);
+  const auto a0 = std::cos(pi * l / 2);
+  const auto a1 = std::sin(pi * l / 2);
+
+  const pcg32 ng0(seeds[0]), ng1(seeds[num_seeds - 1]);
```

Comment:
Why use `pcg32` instead of STL's own?

Reply:
Because I need a PRNG that can be used as a function f(seed, vertex_id). STL's PRNGs can not be used like this. What I do is I seed the PRNG with seed and skip to the vertex_id'th number to make it a function of seed and vertex_id.

Reply:
This ensures that with same seed and same vertex id, any machine rolls the same random number, making LABOR sampling distributed with 0 communication.

Reply:
Not to mention, none of the STL PRNGs are of high quality, I found this out from doing a literature search on this topic before implementing this algorithm, I wouldn't recommend anybody to use them if high quality randomness is needed. They fail many statistical tests, which I need for this algorithm to work properly.

Reply:
> Because I need a PRNG that can be used as a function f(seed, vertex_id) and for different values of seed and vertex_id, rolled randoms need to be uncorrelated, STL doesn't satisfy this requirement. What I do is I seed the PRNG with seed and skip to the vertex_id'th number to make it a function of seed and vertex_id.

A function of seed and vertex_id can also be as simple as like `linear_congruent(seed) ^ vertex_id`.  So would the correlation impact the model performance noticeably?  If not, I feel like we are adding additional technical complexity to the implementation with only theoretical benefit.

> Not to mention, none of the STL PRNGs are of high quality, I found this out from doing a literature search on this topic before implementing this algorithm, I wouldn't recommend anybody to use them if high quality randomness is needed. They fail many statistical tests, which I need for this algorithm to work properly.

Did you observe a gain from using PCG comparing to STL?  If so, by how much?  I'm not quite convinced by introducing a new PRNG solely because of better statistical randomness.

Reply:
I tried doing it your way while I was first developing labor by passing a seed as seed ^ id to STL PRNGs and sampling from it. The results degrade so badly that it looked broken with all numbers significantly different than the cuda implementation.

Reply:
But what I can do now that we don't have the random seed logic, is that I can remove the pcg32 and either use DGL's PRNG or an STL PRNG.

Reply:
> I tried doing it your way while I was first developing labor by passing a seed as seed ^ id to STL PRNGs and sampling from it. The results degrade so badly that it looked broken with all numbers significantly different than the cuda implementation.

This is very intriguing and definitely worth attention.   Do you still have the numbers?  And is the degradation happening on single-machine without cooperative minibatching as well?

I'm not entirely objecting using PCG, especially if not using it will cause a serious breakdown and no longer matches the performance in the paper.  But since it introduces an extra dependency and it requires specific handling (i.e. a hashmap for caching the results), I think it's better to leave a comment in code describing (1) why you prefer PCG over STL (i.e. the performance degradation, preferably with numbers), and (2) what's the issue of using PCG (i.e. slowness, you have to use a hashmap, etc.).

Reply:
The same question goes for having two RNGs with different seeds.  I feel it also needs explanation since it's a non-trivial decision.

Reply:
Sorry for the confusion, the current state of the code has no quality compromise. Using pcg, I could avoid using a hashmap using the same trick in the cuda side but since pcg was slow, I had to reduce the number of calls to it and use a hashmap instead. So using pcg, the code didn't speed up unlike the cuda case. Then, pcg was only needed to get the cooperative minibatching functionality for the cpu side. So I removed it.

The cuda side uses the prng as if it is a function of vertex id and the seed and calls the prng for each edge. Pcg was slow, so I had to cache the random results belonging to particular vertex ids in a hashmap. But if a hashmap is going to be used, then I can use DGL's PRNG too.

Reply:
To clarify, pcg is also needed for heterogenous graphs so that sampling is done globally for the graph with each edge type having the same random seed. Without pcg, this is not possible on the CPU side.